Refactorings

karyon opened the pull request:
see commit messages. also includes #643 with changed commit message because why not.

closes #519

karyon commented:
@diox, the "remove silent exception catching" commit revealed randomly failing test. could you have a look at that? i'm not sure where to start, i can't repro this locally.

here's another test run of the same commit: https://travis-ci.org/django-compressor/django-compressor/builds/80111522

diox commented:
I've restarted the builds that errored but it's failing again. I have no idea why this test fails, I've never seen it fail before...

diox commented:
I'd prefer it this PR was splitted into multiple smaller ones, with tests covering the changes. Otherwise I have no idea what I'm merging or why.

diox commented:
Note: if you rebase, you should benefit from #656 which makes the tests run on travis sudo-less architecture, which means faster builds, which means faster failures to help debugging this :)

karyon commented:
the failures appear to be random. i've restarted the tests about 4 times, and it's always different runs that are failing, but always the same test and error.

you haven't seen this failure before because there was a "except Exception" which hid the problem. i removed the except in a commit.

opening a PR for every single commit seemed a bit overkill to me. but if you really want, i can do that later.

diox commented:
I see. Can you just isolate "remove silent exception catching" in a separate PR ? It'll make it easier to merge the rest first, and then we can concentrate on fixing this random failure.

karyon commented:
alright, i completely separated the commits now. please let me know whether you want anything changed.

diox commented on compressor/base.py:
```diff
@@ -36,17 +36,18 @@ class Compressor(object):
-    type = None
 
-    def __init__(self, content=None, output_prefix=None, context=None, *args, **kwargs):
+    def __init__(self, content=None, output_prefix=None, context=None, filters=None):
```
I prefer leaving `*args` and (more importantly) `**kwargs`, removing them is a backwards-incompatible change for people inheriting from it.

karyon commented:
done.

diox commented:
I'm not a fan of https://github.com/karyon/django-compressor/commit/24551f5dbd28084077eb17b9291b6140fb21de03. It removes methods that are useful for users inheriting from compressor. What was the motivation behind this change ?

karyon commented:
i removed render_cached because there was only one caller and it removing it reduced code complexity noticably. render_output just seemed superfluous so i inlined it, debug_mode was called nowhere so i simply removed it.

i can absolutely restore them if you want, although for render_cached i'd need to check again whether i can keep some of the simplifications, and for debug_mode i'd like to hear when it is intended to be called :)

diox commented:
I kinda like having `render_cached` separated so that it's easy to implement custom behaviour by inheriting from the class (the `is_offline_compression_enabled` exists for a similar purpose, I've used in the the past to make custom compress template tag nodes that are not offline cached, despite behaving normally otherwise).

The rest, and in particular the extra `debug_mode` method, can go.

karyon commented:
i split up the respective commit and added a deprecation warning fix. please review.

basically, the refactoring moves the decision whether to render (cached or offline) from the methods into the main method, which makes things a bit easier to read (for me at least)

karyon commented:
i think https://github.com/karyon/django-compressor/commit/68981ebfbf8af2a47a24a85fcee0d660a556d529#diff-ac8de32c7b59ab69b2ac7137529224f0R85 also kinda includes https://github.com/django-compressor/django-compressor/pull/519.

diox, could you confirm that the offline_manifest[key] cannot return None when key in offline_manifest?

also, since these "refactorings" apparently do fix something, do you think a test is necessary?

diox commented:
`offline_manifest[key]` should not return `None`. It's a simple dict (stored and loaded using a JSON file).

Since you're slightly changing behaviour (for the better), consider adding a test, yes.

karyon commented:
test added.

diox commented:
Excellent, thanks.
